### PR TITLE
Feature: Add Trace to Slf4j MDC

### PR DESCRIPTION
**Pull request by taylor-tian-cn**
This request adds trace information to Slf4j MDC, so they can be inserted into the log text.
This can solve issue #35:Span information in log fields.
Please help review them. Thanks!

### CODE REVIEW

**Comment by pavolloffay** [comment]
hi @euextai thanks for the PR, I think this could be done in https://github.com/opentracing-contrib/java-spring-cloud so another tracer implementations could benefit from it.

**Comment by euextai** [comment]
hi @pavolloffay , I do plan to add this feature in https://github.com/opentracing-contrib/java-spring-cloud. But there are two problems to do so. The first one is that the Tracer is created outside of java-spring-cloud, that means I can't do further more to change Tracer. But in this solution some decorations are needed to create customized ScopeManager and Scope. Another problem is that there's no interface to get the flag about whether the Span is sampled in standard SpanContext. So finally I decided to implement this feature in java-spring-jaeger.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.Map;
+
+import org.slf4j.MDC;
+
+
+/**
+ * @author <a href="mailto:taylor.tian@ericsson.com">Taylor Tian</a>
+ */
+public class ScopeManagerSlf4jDecorator implements ScopeManager {
+  private static final String B3_TRACE_ID = "X-B3-TraceId";
```
Comment (pavolloffay) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Why do you need to distinguish B3 propagation? Cannot we just have two fields `traceid` and `spanid` ?

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeSlf4jDecorator.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+import java.util.Map;
+
+import org.slf4j.MDC;
+
+/**
+ * @author <a href="mailto:taylor.tian@ericsson.com">Taylor Tian</a>
+ */
+public class ScopeSlf4jDecorator implements Scope {
+  private Scope scope;
```
Comment (pavolloffay) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeSlf4jDecorator.java]: make these final

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+  private static final String B3_SPAN_ID = "X-B3-SpanId";
+  private static final String B3_SAMPLED = "X-B3-Sampled";
+  private static final String B3_EXPORT = "X-Span-Export";
+  private static final String OPENTRACING_TRACE_ID = "ToTraceId";
+  private static final String OPENTRACING_SPAN_ID = "ToSpanId";
+  private static final String SAMPLED = "Sampled";
+
+  private ScopeManager scopeManager;
```
Comment (pavolloffay) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: These can be made final

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  }
+
+  @Override
+  public Span activeSpan() {
+    return scopeManager.activeSpan();
+  }
+
+  private Map<String, String> addSpanToMDC(Span span) {
+    JaegerSpanContext context = (JaegerSpanContext) span.context();
+    Map<String, String> currentContext = new HashMap<>();
```
Comment (pavolloffay) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Do we need to create a map? We could put the ids directly into the MDC

**Comment by geoand** [comment]
Thanks for this @euextai! 

I would really like to have some testing for this :)

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 1 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: I want to be compatible with B3 protocol or Spring Sleuth. The log config file won't need to be modified when we migrate legacy system from Sleuth to java-spring-jaeger if we keep compatible with B3.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 4 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Yes, we do. We need to store the context by using a map in order to restore them when a inner span closed.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 3 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: I will amend them.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 1 above)
Comment (pavolloffay) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: I think we should have a separate property to configure this.

One can use B3 with service mesh without using b3 names in the logs.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 4 above)
Comment (pavolloffay) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Alright, then I would initialize it with fixed size and return unmodifiable.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 1 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Actually I use JaegerConfigurationProperties to configure this. Only when 'opentracing.jaeger.enableB3Propagation' is ture, those B3 key will be added to MDC.
`    if (properties.isEnableB3Propagation()) {
      currentContext.put(B3_TRACE_ID, context.toTraceId());
      currentContext.put(B3_SPAN_ID, context.toSpanId());
      currentContext.put(B3_SAMPLED, String.valueOf(context.isSampled()));
      currentContext.put(B3_EXPORT, String.valueOf(context.isSampled()));
    }`

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 4 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Oh, good comments. I will amend it. Thank you.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 1 above)
Comment (pavolloffay) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Yes another property.

Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: OK. I'll add a new properties class Slf4jJaegerProperties to configure these new features.

**Comment by euextai** [comment]
@pavolloffay @geoand if no more comments, I'll amend this commit

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+
+/**
+ * @author <a href="mailto:taylor.tian@ericsson.com">Taylor Tian</a>
+ */
+public class ScopeManagerSlf4jDecorator implements ScopeManager {
+  private static final String B3_TRACE_ID = "X-B3-TraceId";
+  private static final String B3_SPAN_ID = "X-B3-SpanId";
+  private static final String B3_SAMPLED = "X-B3-Sampled";
+  private static final String B3_EXPORT = "X-Span-Export";
+  private static final String OPENTRACING_TRACE_ID = "ToTraceId";
```
Comment (pavolloffay) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: is the `ToTraceId` redered in logs as key for trace id? If so change it to `traceId`. The same for span id

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  public Span activeSpan() {
+    return scopeManager.activeSpan();
+  }
+
+  private Map<String, String> addSpanToMDC(Span span) {
+    JaegerSpanContext context = (JaegerSpanContext) span.context();
+    Map<String, String> currentContext = new HashMap<>();
+    currentContext.put(OPENTRACING_TRACE_ID, context.toTraceId());
+    currentContext.put(OPENTRACING_SPAN_ID, context.toSpanId());
+    currentContext.put(SAMPLED, String.valueOf(context.isSampled()));
```
Comment (pavolloffay) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: We don't want to add context twice if B3 propagation is enabled.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 5 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: yes. I refered to opentracing specification [here](https://github.com/opentracing/specification/blob/master/rfc/trace_identifiers.md). It defines these two ID as ToTraceId & ToSpanId.

Comment (pavolloffay) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Those are method accessor. For the field names I would prefer `traceId` or `trace-id` (w3c trace-context)

Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Oh, I'll change them.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java (hunk 6 above)
Comment (euextai) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeManagerSlf4jDecorator.java]: Ok

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeSlf4jDecorator.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    this.scope = scope;
+    this.previousContext = previousContext;
+  }
+
+  @Override
+  public void close() {
+    scope.close();
+
+    for (Map.Entry<String, String> entry : previousContext.entrySet()) {
+      MDC.put(entry.getKey(), entry.getValue());
```
Comment (GoHahl) [inline opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeSlf4jDecorator.java]: I can't see that you ever remove entries from MDC? So it seems possible that I can have old ID's inside MDC?

**Comment by GoHahl** [comment]
It seems like this approach does not work with Spring WebFlux. Probably because MDC is ThreadLocal if I remember correctly. 
At least for me, there are sometimes some old (incorrect) trace ID's logged.

File: opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeSlf4jDecorator.java (hunk 7 above)
Comment (taylor-tian-cn) [reply opentracing-spring-jaeger-starter/src/main/java/io/opentracing/contrib/java/spring/jaeger/starter/log/slf4j/ScopeSlf4jDecorator.java]: Good comment.  I'll correct it.

**Comment by taylor-tian-cn** [comment]
@GoHahl Thanks for your testing. MDC is ThreadLocal. I'll investigate the solution of WebFlux in Sleuth to see if it is suitable for Jaeger.

**Comment by taylor-tian-cn** [comment]
@GoHahl If ThreadLocal causes MDC problem in WebFlux, I think it will also cause trace context lost in WebFlux. Because default ScopeManager is ThreadLocalScopeManager which is also based on ThreadLocal. So the real problem is about how to integrate Jaeger with webflux, if I was wrong please correct me. In Sleuth, they use TraceWebFilter to keep the trace. So can we open a new issue to track webflux's  integration? I can fix it later. What's your opinions? @pavolloffay @geoand

**Comment by pavolloffay** [comment]
hi @taylor-tian-cn sorry for the delay.

I have just submitted the same feature to a different runtime - Quarkus https://github.com/quarkusio/quarkus/pull/6274. I would rather followed the same approach in this repository it just wraps the existing scope manager and properly clears the context (only added IDs).

Would you like to have a look and make changes?

**Comment by dschulten** [comment]
An alternative to a custom scope manager is the TracerObserver in java-api-extensions. At least I was able to use it. It wraps the tracer with an observable tracer.

**Comment by kfirfer** [comment]
hi, there is something new about this PR ?

**Comment by pavolloffay** [comment]
We have added the feature directly to the jaeger client - https://github.com/jaegertracing/jaeger-client-java/issues/715.

I will close this PR. Could somebody submit  a new PR and use the feature from the Jaeger client?

**Pull request closed without merging**